xwiki-pro-macros-api/src/main/java/com/xwiki/macros/userlist/internal/macro/UserListMacro.java

```diff
@@ -67,22 +74,21 @@ public UserListMacro()
     public List<Block> execute(UserListMacroParameters parameters, String content, MacroTransformationContext context)
         throws MacroExecutionException
     {
-        Template customTemplate = this.templateManager.getTemplate("html_displayer/userreferencelist/view.vm");
 
+        Map<String, String> params = new HashMap<>();
         try {
-            this.bindParameters(parameters);
-            return this.templateManager.execute(customTemplate).getChildren();
-        } catch (Exception e) {
-            throw new MacroExecutionException("Failed to render the userList viewer template.", e);
+            List<String> userReferences =
+                parameters.getUsers().stream().map(reference -> referenceSerializer.serialize(reference)).collect(
+                    Collectors.toList());
+            params.put("users", StringUtils.join(userReferences, ','));
+            params.put("properties", StringUtils.join(parameters.getProperties(), ','));
+            String html = htmlDisplayerManager.display(UserReferenceList.class, content, params, "view");
```

Comment:
The second parameter is the displayed **value**. Isn't the list of user references the displayed value? I don't understand why you pass the ``content`` which should be empty all the time since your macro doesn't support content. I think you should pass directly ``parameters.getUsers()``, and keep the third parameter (``params``) only for stuff that controls how the value (the list of user references) are displayed (i.e. the list of user properties).

Reply:
You're right, the displayed value is the list of user references. I updated the code to pass this value instead of the content. Note that in edit mode, the displayer value is null when rendering `userreferencelist/edit.vm`, I don't understand why. This prevents from editing the user list correctly in the macro dialog. Would you know how to fix this?

Reply:
> Note that in edit mode, the displayer value is null when rendering userreferencelist/edit.vm, I don't understand why. This prevents from editing the user list correctly in the macro dialog.

The Edit Macro (parameters) dialog doesn't call the parameter displayer with the actual parameter value (mostly because when inserting a macro we don't have a value yet), but with the default value (from the parameter descriptor), which in your case is probably null. This way it obtains an edit "template" for the macro parameter, which is used both when inserting a new macro and when editing an existing one. See https://github.com/xwiki-contrib/application-ckeditor/blob/master/application-ckeditor-ui/src/main/resources/CKEditor/MacroService.xml#L280-L282 :

* the edit template is rendered on the server side using the default value
* the actual value is set on the client side

So what is exactly the problem? The user picker is not initialized (on the client side) with the macro parameter value when editing an existing macro? Either the edit (HTML) "template" returned from the server side is not right or the code that sets the value on the client side has a bug / limitation. But I think we have other macros that use a picker (e.g. office macro) and for which editing an existing macro works fine.

Reply:
I just tested and editing an existing office macro call works as expected (the value is properly set on the picker).